### PR TITLE
[RO-264] Enhance anemic assignment entity

### PR DESCRIPTION
**Pull request by Skyld**
Ticket: https://oat-sa.atlassian.net/browse/RO-264

Main purpose of this PR is to transform `Assignment` entity from anemic to rich object model. Rest of the entities will come later in another PRs. What was also changed:

### Added
- Added `status` column to `assignments` database table.

### Removed
- Removed `state` column from `assignments` database table.

### CODE REVIEW

File: src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php [hunk 1]
```diff
@@ -129,14 +128,14 @@ private function processOperation(BulkOperation $operation, BulkResult $result):
     /**
      * @throws LogicException
      */
-    private function validateStateTransition(BulkOperation $operation): void
+    private function validateStateAttribute(BulkOperation $operation): void
     {
-        if ($operation->getAttribute('state') !== Assignment::STATE_CANCELLED) {
+        if ($operation->getAttribute('state') !== Assignment::STATUS_CANCELLED) {
```
Comment (andreluizmachado) [inline src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php:133]: can we change it to `$operation->getAttribute('status')` as well?

File: src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php [hunk 2]
```diff
@@ -129,14 +128,14 @@ private function processOperation(BulkOperation $operation, BulkResult $result):
     /**
      * @throws LogicException
      */
-    private function validateStateTransition(BulkOperation $operation): void
+    private function validateStateAttribute(BulkOperation $operation): void
```
Comment (andreluizmachado) [inline src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php:131]: can we change it to `validateStatusAttribute` as well?

File: tests/Fixtures/userWithAllAttemptsTaken.yml [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
         groupId: 'group_1'
         assignments: [ '@assignment2' ]
+
+OAT\SimpleRoster\Entity\Assignment:
+    assignment2:
+        __construct:
+            id: <uuidV6(00000003-0000-6000-0000-000000000000)>
+            status: 'ready'
+            lineItem: '@lineItem2'
+            attemptsCount: 2
```
Comment (andreluizmachado) [inline tests/Fixtures/userWithAllAttemptsTaken.yml:26]: missing empty line

File: tests/Fixtures/userWithReadyAssignment.yml [hunk 4]
```diff
@@ -29,7 +29,8 @@ OAT\SimpleRoster\Entity\User:
...
     assignment1:
-        id: <uuidV6('00000001-0000-6000-0000-000000000000')>
-        lineItem: '@lineItem1'
-        state: 'ready'
-        attemptsCount: 1
+        __construct:
+            id: <uuidV6('00000001-0000-6000-0000-000000000000')>
+            status: 'ready'
+            lineItem: '@lineItem1'
+            attemptsCount: 1
```
Comment (andreluizmachado) [inline tests/Fixtures/userWithReadyAssignment.yml:36]: missing empty line

File: tests/Integration/Entity/UserTest.php [hunk 5]
```diff
@@ -60,4 +61,14 @@ public function testItCanRetrieveAndRemoveAssignments(): void
         self::assertEmpty($subject->getAssignments());
         self::assertEmpty($subject->getAvailableAssignments());
     }
+
+    public function testItThrowsExceptionIfUserHasNoAssignments(): void
+    {
+        $this->expectException(AssignmentNotFoundException::class);
+        $this->expectExceptionMessage("User 'testUser' does not have any assignments.");
+
+        (new User())
```
Comment (andreluizmachado) [inline tests/Integration/Entity/UserTest.php:70]: You will have to change it after this [PR](https://github.com/oat-sa/simple-roster/pull/211).

**Review by andreluizmachado — CHANGES_REQUESTED** [review]

File: src/Entity/Assignment.php [hunk 6]
```diff
@@ -24,76 +24,87 @@
...
 
-    public function getState(): ?string
+    public function getId(): UuidV6
     {
-        return $this->state;
+        return $this->id;
     }
 
-    public function setState(string $state): self
+    public function getStatus(): ?string
```
[the remark this answers is not in the document]
Comment (Skyld) [reply src/Entity/Assignment.php:94]: Totally, fixed!

File: src/Entity/Assignment.php [hunk 7]
```diff
@@ -121,74 +132,132 @@ public function getUpdatedAt(): ?DateTime
...
+    {
+        return in_array($this->status, [self::STATUS_STARTED, self::STATUS_READY], true);
+    }
+
+    /**
+     * @throws InvalidAssignmentStatusTransitionException
+     */
     public function complete(): self
     {
+        if ($this->status !== self::STATUS_STARTED) {
```
[the remark this answers is not in the document]
Comment (Skyld) [reply src/Entity/Assignment.php:219]: No, only started ones.

File: src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php [hunk 8]
```diff
@@ -28,15 +28,14 @@
...
-use OAT\SimpleRoster\Bulk\Processor\BulkOperationCollectionProcessorInterface;
 use OAT\SimpleRoster\Bulk\Result\BulkResult;
 use OAT\SimpleRoster\Entity\Assignment;
 use OAT\SimpleRoster\Entity\User;
 use OAT\SimpleRoster\Repository\UserRepository;
 use Psr\Log\LoggerInterface;
 use Throwable;
 
-class BulkUpdateUsersAssignmentsStateService implements BulkOperationCollectionProcessorInterface
+class BulkUpdateUsersAssignmentsStateService
```
[the remark this answers is not in the document]
Comment (Skyld) [reply src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php:38]: Fixed

File: src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php (hunk 1 above)
Comment (Skyld) [reply src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php:133]: We can, I am going to create a PR in lambda assignment manager as well to match expectations. Fixed

File: src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php (hunk 2 above)
Comment (Skyld) [reply src/Service/Bulk/BulkUpdateUsersAssignmentsStateService.php:131]: Fixed

File: tests/Fixtures/userWithAllAttemptsTaken.yml (hunk 3 above)
Comment (Skyld) [reply tests/Fixtures/userWithAllAttemptsTaken.yml:26]: Fixed

File: tests/Integration/Entity/UserTest.php (hunk 5 above)
Comment (Skyld) [reply tests/Integration/Entity/UserTest.php:70]: Sure 👍

**Review by andreluizmachado — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Changelog is updated according to changes (if applicable)
- [x] Documentation is updated according to changes (if applicable)

**Review by luisfn — APPROVED** [review]
**Review Checklist**
- [X] New code is covered by tests (if applicable)
- [X] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [X] Changelog is updated according to changes (if applicable)
- [X] Documentation is updated according to changes (if applicable)

**Pull request merged**